composition-js/src/merging/merge.ts

```diff
@@ -1102,40 +1106,55 @@ class Merger {
   }
 
   private addArgumentsShallow<T extends FieldDefinition<any> | DirectiveDefinition>(sources: (T | undefined)[], dest: T) {
+    const argNames = new Set<string>();
     for (const source of sources) {
       if (!source) {
         continue;
       }
-      for (const argument of source.arguments()) {
-        if (!dest.argument(argument.name)) {
-          dest.addArgument(argument.name);
-        }
-      }
-    }
-  }
-
-  private mergeArgument(sources: (ArgumentDefinition<any> | undefined)[], dest: ArgumentDefinition<any>, useIntersection: boolean = false) {
-    if (useIntersection) {
-      for (const source of sources) {
-        if (!source) {
+      source.arguments().forEach((arg) => argNames.add(arg.name));
+    }
+
+    for (const argName of argNames) {
+      // We add the arguement unconditionally even if we're going to remove it in
+      // some path. Done because this helps reusing our "reportMismatchHint" method
+      // in those cases.
+      const arg = dest.addArgument(argName);
+      // If all the sources that have the field have the argument, we do merge it
+      // and we're good, but otherwise ...
+      if (!sources.every((s) => !s || s.argument(argName))) {
+        // ... we don't merge the argument: some subgraphs wouldn't know what
+        // to make of it and that would be dodgy at best. If the argument is
+        // optional in all sources, then we can compose properly and just issue a
+        // hint. But if it is mandatory, then we have to fail composition, otherwise
+        // the query planner would have no choice but to generate invalidate queries.
+        const nonOptionalSources = sources.map((s, i) => s && s.argument(argName)?.isRequired() ? this.names[i] : undefined).filter((s) => !!s) as string[];
+        if (nonOptionalSources.length > 0) {
```

Comment:
**question/nitpick** Do we normally compare against `> 0`?

```suggestion
        if (nonOptionalSources.length) {
```

Reply:
This seems fine to me, I don't feel strongly. I just wanna make sure I stay consistent when I do this sort of thing.

(Full disclosure: I'm normally in the "don't compare against 0" camp but 🤷 )